parquet/src/main/java/org/apache/iceberg/parquet/ParquetVariantReaders.java

```diff
@@ -0,0 +1,409 @@
+      return reader.column();
+    }
+
+    @Override
+    public List<TripleIterator<?>> columns() {
+      return reader.columns();
+    }
+
+    @Override
+    public void setPageSource(PageReadStore pageStore) {
+      reader.setPageSource(pageStore);
+    }
+  }
+
+  private static ByteBuffer readBinary(ColumnIterator<?> column) {
+    ByteBuffer data = column.nextBinary().toByteBuffer();
+    byte[] array = new byte[data.remaining()];
+    data.get(array, 0, data.remaining());
+    return ByteBuffer.wrap(array).order(ByteOrder.LITTLE_ENDIAN);
+  }
+
+  private static class VariantMetadataReader extends PrimitiveReader<VariantMetadata> {
+    private VariantMetadataReader(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public VariantMetadata read(VariantMetadata reuse) {
+      return Variants.metadata(readBinary(column));
+    }
+  }
+
+  private static class SerializedVariantReader extends PrimitiveReader<VariantValue>
+      implements VariantValueReader {
+    private SerializedVariantReader(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public VariantValue read(VariantMetadata metadata) {
+      return Variants.value(metadata, readBinary(column));
+    }
+  }
+
+  private static class ValueAsVariantReader<V> extends DelegatingValueReader<V, VariantValue>
+      implements VariantValueReader {
+    private final PhysicalType type;
+
+    private ValueAsVariantReader(PhysicalType type, ParquetValueReader<V> reader) {
+      super(reader);
+      this.type = type;
+    }
+
+    @Override
+    public VariantValue read(VariantMetadata ignored) {
+      return Variants.of(type, readFromDelegate(null));
+    }
+  }
+
+  /**
+   * A Variant reader that combines value and typed_value columns from Parquet.
+   *
+   * <p>This reader does not handle merging partially shredded objects. To handle shredded objects,
+   * use {@link ShreddedObjectReader}.
+   */
+  private static class ShreddedVariantReader implements VariantValueReader {
+    private final int valueDL;
+    private final VariantValueReader valueReader;
+    private final int typedDL;
+    private final VariantValueReader typedReader;
+    private final TripleIterator<?> column;
+    private final List<TripleIterator<?>> children;
+
+    private ShreddedVariantReader(
+        int valueDL, VariantValueReader valueReader, int typedDL, VariantValueReader typedReader) {
+      this.valueDL = valueDL;
+      this.valueReader = valueReader;
+      this.typedDL = typedDL;
+      this.typedReader = typedReader;
+      this.column = valueReader != null ? valueReader.column() : typedReader.column();
+      this.children = children(valueReader, typedReader);
+    }
+
+    @Override
+    public VariantValue read(VariantMetadata metadata) {
+      VariantValue value = ParquetVariantReaders.read(metadata, valueReader, valueDL);
+      VariantValue typed = ParquetVariantReaders.read(metadata, typedReader, typedDL);
+
+      if (typed != null) {
```

Comment:
I don't recall where the spec landed on this, but shouldn't we fail if `typed` and `value` are both non-null?

Reply:
Yes, I'll update this to implement the latest version of the spec.

Reply:
Fixed.